File: solutions/elixir/tournament/2/lib/tournament.ex
defmodule Tournament do
  @doc """
  Given `input` lines representing two teams and whether the first of them won,
  lost, or reached a draw, separated by semicolons, calculate the statistics
  for each team's number of games played, won, drawn, lost, and total points
  for the season, and return a nicely-formatted string table.

  A win earns a team 3 points, a draw earns 1 point, and a loss earns nothing.

  Order the outcome by most total points for the season, and settle ties by
  listing the teams in alphabetical order.
  """

  @rules %{
    "win" => %{mp: 1, w: 1, p: 3},
    "loss" => %{mp: 1, l: 1},
    "draw" => %{mp: 1, d: 1, p: 1}
  }

  @spec tally(input :: list(String.t())) :: String.t()
  def tally(input) do
    output = input
    |> Enum.filter(&valid?/1)
    |> Enum.map(fn item -> String.split(item, ";") end)
    |> Enum.reduce(%{}, &calculate_stats/2)
    |> Map.to_list
    |> Enum.sort(fn({t1, %{p: a}}, {t2, %{p: b}}) ->
      cond do
        a == b -> t1 <= t2
        true -> a >= b
      end
    end)
    |> Enum.map(&format_output/1)
    |> Enum.join("\n")

    if output == "" do
      header()
    else
      ~s(#{header()}\n#{output})
    end
  end

  defp valid?(line) do
    parts = String.split(line, ";")
    if length(parts) == 3 do
      [_, _, outcome] = parts
      Enum.member?(["win", "loss", "draw"], outcome)
    else
      false
    end
  end

  defp calculate_stats([team1, team2, outcome], acc) do
    acc
    |> Map.put(team1, do_stats(acc, team1, 0, outcome))
    |> Map.put(team2, do_stats(acc, team2, 1, outcome))
  end

  defp do_stats(acc, team, position, outcome) do
    team
    |> get_stats(acc)
    |> update_stats(position, outcome)
  end

  defp get_stats(team, acc) do
    Map.get(acc, team, %{mp: 0, w: 0, l: 0, d: 0, p: 0})
  end

  defp update_stats(stats, 0, "win"), do: apply_rules(stats, "win")
  defp update_stats(stats, 1, "loss"), do: apply_rules(stats, "win")
  defp update_stats(stats, 0, "loss"), do: apply_rules(stats, "loss")
  defp update_stats(stats, 1, "win"), do: apply_rules(stats, "loss")
  defp update_stats(stats, _, "draw"), do: apply_rules(stats, "draw")

  defp apply_rules(stats, outcome) do
    Map.merge(stats, @rules[outcome], fn (_k, v1, v2) -> v1 + v2 end)
  end

  defp header() do
    """
    Team                           | MP |  W |  D |  L |  P
    """ |> String.trim
  end

  defp format_output({team, stats}) do
    scores = [:mp, :w, :d, :l, :p]
    |> Enum.map(fn key ->
      " | #{String.pad_leading(to_string(Map.get(stats, key)), 2)}"
    end)
    "#{String.pad_trailing(team, 30)}#{scores}"
  end
end

File: solutions/typescript/tournament/1/tournament.ts
export class Tournament {
  
  public tally(input: string): string {
    const lines = input.split('\n');
    const teams: { [key: string]: { points: number; matches: number, wins: number, losses: number, draws: number } } = {};

    for (const line of lines) {
      const [team1, team2, outcome] = line.split(';');
      if (!team1 || !team2 || !outcome) {
        continue; // Skip invalid lines
      }
      const points1 = this.getPoints(outcome === 'win', outcome === 'loss');
      const points2 = this.getPoints(outcome === 'loss', outcome === 'win');

      teams[team1] = teams[team1] || { points: 0, matches: 0, wins: 0, losses: 0, draws: 0 };
      teams[team2] = teams[team2] || { points: 0, matches: 0, wins: 0, losses: 0, draws: 0 };

      teams[team1].points += points1;
      teams[team2].points += points2;
      teams[team1].matches += 1;
      teams[team2].matches += 1;

      teams[team1].wins += points1 === 3 ? 1 : 0;
      teams[team2].wins += points2 === 3 ? 1 : 0;
      teams[team1].losses += points1 === 0 ? 1 : 0;
      teams[team2].losses += points2 === 0 ? 1 : 0;
      teams[team1].draws += points1 === 1 ? 1 : 0;
      teams[team2].draws += points2 === 1 ? 1 : 0;
    }

    console.log(teams);

    const sortedTeams = Object.entries(teams)
      .sort((a, b) => b[1].points - a[1].points || a[0].localeCompare(b[0]));

    let results = sortedTeams.map(([team, { matches, points }]) => {
      const wins = teams[team].wins || 0;
      const draws = teams[team].draws || 0;
      const losses = matches - wins - draws;
      return `${team.padEnd(30)} | ${matches.toString().padStart(2)} | ${wins.toFixed(0).padStart(2)} | ${draws.toString().padStart(2)} | ${losses.toString().padStart(2)} | ${points.toString().padStart(2)  }`;
    });
    results.unshift('Team                           | MP |  W |  D |  L |  P');

    return results.join('\n');
  }

  private getPoints(win: boolean, loss: boolean): number {
    if (win) return 3;
    if (loss) return 0;
    return 1; // For a draw
  }
}
